Sources/SwiftGRPCNIO/HTTP1ToRawGRPCServerCodec.swift

```diff
@@ -27,111 +27,200 @@ public enum RawGRPCServerResponsePart {
+  struct StateMachineError: Error, GRPCStatusTransformable {
+    private let message: String
+
+    init(_ message: String) {
+      self.message = message
+    }
+
+    func asGRPCStatus() -> GRPCStatus {
+      return GRPCStatus.processingError
+    }
+  }
 }
 
 extension HTTP1ToRawGRPCServerCodec: ChannelInboundHandler {
   public typealias InboundIn = HTTPServerRequestPart
   public typealias InboundOut = RawGRPCServerRequestPart
-  
+
   public func channelRead(ctx: ChannelHandlerContext, data: NIOAny) {
-    switch self.unwrapInboundIn(data) {
-    case .head(let requestHead):
-      guard case .expectingHeaders = state
-        else { preconditionFailure("received headers while in state \(state)") }
-
-      state = .expectingCompressedFlag
-      buffer = ctx.channel.allocator.buffer(capacity: 5)
-
-      ctx.fireChannelRead(self.wrapInboundOut(.head(requestHead)))
-
-    case .body(var body):
-      guard var buffer = buffer
-        else { preconditionFailure("buffer not initialized") }
-      assert(state.expectingBody, "received body while in state \(state)")
-      buffer.write(buffer: &body)
-
-      // Iterate over all available incoming data, trying to read length-delimited messages.
-      // Each message has the following format:
-      // - 1 byte "compressed" flag (currently always zero, as we do not support for compression)
-      // - 4 byte signed-integer payload length (N)
-      // - N bytes payload (normally a valid wire-format protocol buffer)
-      requestProcessing: while true {
-        switch state {
-        case .expectingHeaders: preconditionFailure("unexpected state \(state)")
-        case .expectingCompressedFlag:
-          guard let compressionFlag: Int8 = buffer.readInteger() else { break requestProcessing }
-          //! FIXME: Avoid crashing here and instead drop the connection.
-          precondition(compressionFlag == 0, "unexpected compression flag \(compressionFlag); compression is not supported and we did not indicate support for it")
-          state = .expectingMessageLength
-
-        case .expectingMessageLength:
-          guard let messageLength: UInt32 = buffer.readInteger() else { break requestProcessing }
-          state = .receivedMessageLength(messageLength)
-
-        case .receivedMessageLength(let messageLength):
-          guard let messageBytes = buffer.readBytes(length: numericCast(messageLength)) else { break }
-
-          //! FIXME: Use a slice of this buffer instead of copying to a new buffer.
-          var responseBuffer = ctx.channel.allocator.buffer(capacity: messageBytes.count)
-          responseBuffer.write(bytes: messageBytes)
-          ctx.fireChannelRead(self.wrapInboundOut(.message(responseBuffer)))
-          //! FIXME: Call buffer.discardReadBytes() here?
-          //! ALTERNATIVE: Check if the buffer has no further data right now, then clear it.
-
-          state = .expectingCompressedFlag
-        }
+    if case .ignore = inboundState { return }
```

Comment:
Would it be worth logging this (or sending it to an error delegate/calling `ctx.fireErrorCaught`) if hit? (Maybe not, given that something else would probably have gone wrong before this could even happen.)

Reply:
I keep going back and forth on this. I don't feel strongly either way so will leave it for now. 

You're right in that this will happen if something's gone wrong or we've sent the status. In either case sending the status should close the stream, so we should stop receiving messages at that point anyway.